### PR TITLE
Unit test report tolerance at metric level

### PR DESCRIPTION
**Pull request by vikash018**
Unit test report tolerance at metric level

### CODE REVIEW

File: verticapy/tests_new/machine_learning/vertica/__init__.py [hunk 1]
```diff
@@ -17,48 +17,497 @@
...
+rel_abs_tol_map = {
+    "RandomForestRegressor": {
+        "explained_variance": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+        "max_error": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "median_absolute_error": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+        "mean_absolute_error": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+        "mean_squared_error": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+        "mean_squared_log_error": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+        "rmse": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+        **dict.fromkeys(["r2", "R-squared"], {"rel": 1e-01, "abs": ABS_TOLERANCE}),
```
Comment (mail4umar) [inline verticapy/tests_new/machine_learning/vertica/__init__.py:29]: I think we should be as tight as possible on these R-squared values. Did you try 0.05? I think with the same data, we should be in that range.

File: verticapy/tests_new/machine_learning/vertica/__init__.py [hunk 2]
```diff
@@ -17,48 +17,497 @@
...
+        "prc_auc": {"rel": 1e-02, "abs": ABS_TOLERANCE},
+        "accuracy": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+        "log_loss": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "precision": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+        "recall": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "f1_score": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "mcc": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+        "markedness": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "csi": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+        "confusion_matrix": {"rel": 1e-00, "abs": ABS_TOLERANCE},
```
Comment (mail4umar) [inline verticapy/tests_new/machine_learning/vertica/__init__.py:63]: Are you comparing all the 4 parts of the matrix: TP, FP, TN, FN?

File: verticapy/tests_new/machine_learning/vertica/__init__.py [hunk 3]
```diff
@@ -17,48 +17,497 @@
...
+    },
+    "DecisionTreeRegressor": {
+        "explained_variance": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "max_error": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+        "median_absolute_error": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+        "mean_absolute_error": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+        "mean_squared_error": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "mean_squared_log_error": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "rmse": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+        **dict.fromkeys(["r2", "R-squared"], {"rel": 1e-00, "abs": ABS_TOLERANCE}),
```
Comment (mail4umar) [inline verticapy/tests_new/machine_learning/vertica/__init__.py:82]: R-squared tolerance of 1? That should not be the case. It is too huge.

File: verticapy/tests_new/machine_learning/vertica/__init__.py [hunk 4]
```diff
@@ -17,48 +17,497 @@
...
+        "load_model": {"rel": 1e-02, "abs": ABS_TOLERANCE},
+    },
+    "DecisionTreeClassifier": {
+        "auc": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "prc_auc": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "accuracy": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "log_loss": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "precision": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "recall": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "f1_score": {"rel": 1e-00, "abs": ABS_TOLERANCE},
```
Comment (mail4umar) [inline verticapy/tests_new/machine_learning/vertica/__init__.py:112]: F-score ranges from 0 to 1. So this is again too high.

File: verticapy/tests_new/machine_learning/vertica/__init__.py [hunk 5]
```diff
@@ -17,48 +17,497 @@
...
+    },
+    "XGBRegressor": {
+        "explained_variance": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "max_error": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "median_absolute_error": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "mean_absolute_error": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "mean_squared_error": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "mean_squared_log_error": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "rmse": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        **dict.fromkeys(["r2", "R-squared"], {"rel": 1e-00, "abs": ABS_TOLERANCE}),
```
Comment (mail4umar) [inline verticapy/tests_new/machine_learning/vertica/__init__.py:190]: R-squared tolerance too high.

File: verticapy/tests_new/machine_learning/vertica/__init__.py [hunk 6]
```diff
@@ -17,48 +17,497 @@
...
+        "load_model": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+    },
+    "XGBClassifier": {
+        "auc": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+        "prc_auc": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+        "accuracy": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+        "log_loss": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "precision": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "recall": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "f1_score": {"rel": 1e-00, "abs": ABS_TOLERANCE},
```
Comment (mail4umar) [inline verticapy/tests_new/machine_learning/vertica/__init__.py:221]: F1 score tolerance too high

File: verticapy/tests_new/machine_learning/vertica/__init__.py [hunk 7]
```diff
@@ -17,48 +17,497 @@
...
+    },
+    "MA": {
+        "explained_variance": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+        "max_error": {"rel": 1e-00, "abs": ABS_TOLERANCE},
+        "median_absolute_error": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+        "mean_absolute_error": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+        "mean_squared_error": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+        "mean_squared_log_error": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+        "rmse": {"rel": 1e-01, "abs": ABS_TOLERANCE},
+        **dict.fromkeys(["r2", "R-squared"], {"rel": 1e-01, "abs": ABS_TOLERANCE}),
```
Comment (mail4umar) [inline verticapy/tests_new/machine_learning/vertica/__init__.py:460]: Why does ARMA have a much looser tolerance than AR or ARMA? Did you try tighter?

**Review by mail4umar — CHANGES_REQUESTED** [review]

File: verticapy/tests_new/machine_learning/vertica/__init__.py (hunk 1 above)
Comment (vikash018) [reply verticapy/tests_new/machine_learning/vertica/__init__.py]: As discussed, I have added code to calculate tolerance using
_rel_tol = abs(vpy_score - py_score) / min(abs(vpy_score), abs(py_score))
_abs_tol = abs(vpy_score - py_score) / (1 + min(abs(vpy_score), abs(py_score)))
and accordingly updated value in __init__.py file.

File: verticapy/tests_new/machine_learning/vertica/__init__.py (hunk 2 above)
Comment (vikash018) [reply verticapy/tests_new/machine_learning/vertica/__init__.py]: yes.

File: verticapy/tests_new/machine_learning/vertica/__init__.py (hunk 3 above)
Comment (vikash018) [reply verticapy/tests_new/machine_learning/vertica/__init__.py]: same as first comment.

File: verticapy/tests_new/machine_learning/vertica/__init__.py (hunk 4 above)
Comment (vikash018) [reply verticapy/tests_new/machine_learning/vertica/__init__.py]: same as first comment.

File: verticapy/tests_new/machine_learning/vertica/__init__.py (hunk 6 above)
Comment (vikash018) [reply verticapy/tests_new/machine_learning/vertica/__init__.py]: same as first comment.

File: verticapy/tests_new/machine_learning/vertica/__init__.py (hunk 7 above)
Comment (vikash018) [reply verticapy/tests_new/machine_learning/vertica/__init__.py]: same as first comment.

File: verticapy/tests_new/machine_learning/vertica/__init__.py (hunk 5 above)
Comment (vikash018) [reply verticapy/tests_new/machine_learning/vertica/__init__.py]: same as first comment.

**Pull request merged**